### ISSUE TITLE
providing more distribution-friendly artefacts for releases

### ISSUE THREAD

**Issue by cyphar** [trezor/trezord-go#46]
I'm planning on packaging this project for openSUSE (so users don't need to install third-party RPMs). From the perspective of a distribution, the most safe way for a project to provide source code that a distribution can use for building is to provide:

* A deterministic `.tar.gz` of *all* the source code, including any vendored dependencies in Go. This can be generated from `git` pretty easily.
* A detached signature of the above `.tar.gz`.
* A signed set of hashes of the built binaries would also be useful for reproducible build checking, if the project is reproducibly built (which it is since it's a Go project).
* And a link to the PGP key used for the above signing -- preferably in an `gpg-offline` format but that's not required.

(Though you should note that we don't *need* any of this, I can already make a package of `trezord` with the existing git tags.)

I already have a script that does all of this, and is used by both [`runc`](https://github.com/opencontainers/runc/blob/v1.0.0-rc5/script/release.sh) (the runtime that Docker uses) as well as [`umoci`](https://github.com/openSUSE/umoci/blob/v0.4.0/hack/release.sh). This script is written so that you can easily swap out the distribution-specific parts of the build process, but still have the same structure of files to upload.

If you like I can submit a PR which will include the above script so that you can publish these files as well. If you want I can also make this a separate issue, but this felt to the most appropriate place to put this comment.

**Comment by prusnak** [comment]
Release tarballs are available from here: https://github.com/trezor/trezord-go/releases (they are created automagically by github when there is a new tag added).

Commits by karelb and me are signed, so there is already a signed content.

I would rather not see distribution specific stuff in the repo. But you can help us with other stuff you mentioned that is usable for everyone (signed set of hashes, etc).

**Comment by cyphar** [comment]
> Release tarballs are available from here: https://github.com/trezor/trezord-go/releases (they are created automagically by github when there is a new tag added).

Yes, but if you want to publish a signed set of hashes including the archive (or detached signatures of the archive) then you would have to download them from GitHub and then re-upload the signed hashes. While possible, it's more effort (for you) than having a script which generates the archives and the hashes and the signatures (I wrote the scripts for projects I maintain specifically to avoid having to deal with this stuff).

There's also a slight benefit over GitHub's archives (the version is included as the directory name in the archive, which matches standards for source archives better -- but that's mostly a taste thing). Also the contents of the archive can be reproducibly generated from a commit (I don't know what command GitHub uses to generate their archives because they don't tell us).

> Commits by karelb and me are signed, so there is already a signed content.

Yes, and I've seen that you also sign tags -- this is much better than most projects I've seen! However, most distributions have tooling around signed tarballs rather than signed git tags (usually build systems take the tarball rather than a pointer to the source code). To be fair, this is a bit of a "distributions are weird" thing, but it also means that we don't need to store the entire git tree in the source packages and users can still verify the signatures.

> I would rather not see distribution specific stuff in the repo.

I completely agre -- this scripts I pointed to aren't distributions-specific at all.

> But you can help us with other stuff you mentioned that is usable for everyone (signed set of hashes, etc).

Cool, I'll send a PR when I have a chance. :D

**Comment by cyphar** [comment]
FWIW, I've [packaged `trezord` for openSUSE](https://build.opensuse.org/package/show/network:cryptocurrencies/trezord). I will make a PR to improve the artefact situation as soon as I get a chance. :D
